lib/backend/postgres/driver.go

```diff
@@ -0,0 +1,233 @@
+	err = pgConn.QueryRow(ctx, query, connConfig.Database).Scan(&exists)
+	if err != nil {
+		return trace.BadParameter("failed to verify %q database exists: %v", connConfig.Database, err)
+	}
+
+	// Create database if it doesn't exist.
+	if !exists {
+		_, err = pgConn.Exec(ctx, fmt.Sprintf("CREATE DATABASE %v", connConfig.Database))
+		if err != nil {
+			return trace.BadParameter("failed to create %q database: %v", connConfig.Database, err)
+		}
+		d.cfg.Log.Infof("Created storage database %q", connConfig.Database)
+	}
+
+	return nil
+}
+
+// url returns a connection string URL created from pgDriver's config.
+func (d *pgDriver) url() *url.URL {
+	u := url.URL{
+		Scheme: "postgres",
+		Host:   d.cfg.Addr,
+		Path:   "/" + d.cfg.Database,
+	}
+	q := u.Query()
+	q.Set("sslmode", "verify-full")
+	q.Set("sslrootcert", d.cfg.TLS.CAFile)
+	q.Set("sslcert", d.cfg.TLS.ClientCertFile)
+	q.Set("sslkey", d.cfg.TLS.ClientKeyFile)
+	u.RawQuery = q.Encode()
+	return &u
+}
+
+// pgDB implements sqlbk.DB. It is returned from pgDriver.open.
+type pgDB struct {
+	*sql.DB
+	*pgDriver
+	readOnlyOpts  *sql.TxOptions
+	readWriteOpts *sql.TxOptions
+}
+
+// Begin a read/write transaction.
+func (db *pgDB) Begin(ctx context.Context) sqlbk.Tx {
+	return db.begin(ctx, db.readWriteOpts)
+}
+
+// ReadOnly begins a read-only transaction. Calling a mutating Tx method
+// will result in a failed transaction.
+func (db *pgDB) ReadOnly(ctx context.Context) sqlbk.Tx {
+	return db.begin(ctx, db.readOnlyOpts)
+}
+
+// begin a transaction with options (read/write or read-only).
+func (db *pgDB) begin(ctx context.Context, opts *sql.TxOptions) *pgTx {
+	tx, err := db.DB.BeginTx(ctx, opts)
+	return &pgTx{
+		opts:  opts,
+		pgDB:  db,
+		sqlTx: tx,
+		ctx:   ctx,
+		err:   convertError(err),
+	}
+}
+
+// sqlNullTime converts a time to a nullable sql time, which is required when
+// passing time parameters for nullable SQL database columns such as expires.
+func sqlNullTime(t time.Time) sql.NullTime {
+	if t.IsZero() {
+		return sql.NullTime{}
+	}
+	return sql.NullTime{Time: t, Valid: true}
+}
+
+// newID returns a new backend item ID. A backend item's ID is unique per key.
+//
+// It returns the current UnixNano time. A clockwork.Clock is not used here
+// because it would not be unique for tests using a fake clock. The number
+// returned can be anything that has a high probability of being unique per key
+// and is incremental.
+func newID() int64 {
+	return time.Now().UnixNano()
+}
+
+// convertError to a trace.Error.
+func convertError(err error) error {
+	if err == nil {
+		return nil
+	}
+	if errors.Is(err, sql.ErrNoRows) {
+		return trace.NotFound(err.Error())
+	}
+	if pgErr, ok := err.(*pgconn.PgError); ok {
+		switch pgErr.Code {
+		case errCodeUniqueConstraint:
+			return trace.AlreadyExists("already exists")
+		case errCodeNotSerializable:
+			return sqlbk.ErrRetry
+		}
+	}
+	return trace.Wrap(err)
+}
```

Comment:
The use of `NotFound` and `AlreadyExists` in a generically applied conversion function makes me a bit nervous, since these errors coming from a backend have special meaning.  In particular, we've previously seen bugs where backends emitted `NotFound` errors in cases where the underlying error meant something other than "item at requested key does not exist", and it would cause the cache to enter a permanently bad state (IIRC it caused a fairly serious outage actually).

While I haven't noticed any particular issues in the code as-is, I'm concerned that a blanket conversion like this greatly increases the likelihood that a bug of this nature would get missed during reviews of future changes, and would be more difficult to discover during debugging.

I think this risk is further compounded by the fact that a lot of error propagation and error-related control-flow in this PR seems to happen via hidden state, making it less obvious how errors are handled, and increasing the cognitive load of reviewing the unhappy path (this conversion function would likely not even appear in most future diffs/PRs related to this backend, despite its ubiquity).

This kind of thing is a bit subjective so take it or leave it, but I feel that we'd benefit in the long run from moving to explicit conversions at the query-site.  Especially in light of the non-standard error propagation path.

Reply:
I thought the convert error function was a standard Teleport thing since it exists in both lite and dynamo backends. Although the lite version is a challenge to understand. Such a function was not in my plans but it turned out to be useful in the end.

I share your concern with returning the wrong type of error. Much of getting the backend compliance suite to complete successfully involved ensuring the correct error was returned. But I think saying the error handing in this PR is hidden or non-standard is a bit much.

Most of the error handling that matters is in Backend methods, e.g. Create returns AlreadyExists and CompareAndSwap returns CompareFailed. I wanted to keep all such error handling in Backend so it would not be a concern when new drivers are created. But there is a need to convert platform-specific error codes to something the Backend understands. In the end, only 3 types of errors were needed or 2 types per type of query. Serializable errors can occur with any query. SELECT queries may return a no rows found error (NotFound), and INSERT/UPDATE queries may return a unique constraint violation (AlreadyExists). SELECT queries are called from Get Backend methods and INSERT/UPDATE from Create/Put/Update/KeepAlive. 

Perhaps the real issue is that the errors in question pass through the Backend. Maybe if drivers converted errors to some other signal meant only for the Backend then the Backend methods could explicitly trap the error and convert it to the correct target. E.g. drivers return BadParameter(“not found”) and Backend Get methods trap and return NotFound. Then all backend methods would return errors explicitly.

Otherwise, I’ll think about whether a backend method could return the wrong error and/or how to change the design to ensure NotFound is not returned from a Put method in the future by some rogue non-standard SQL engine. I’m not so certain inlining rollback and convert functions would be an improvement, but I’ll continue to think about it as well.

Reply:
> I thought the convert error function was a standard Teleport thing since it exists in both lite and dynamo backends.

Having wrestled with bugs due to dynamo error types is a big part of why I'm sensitive to this kind of thing :D

> Maybe if drivers converted errors to some other signal meant only for the Backend then the Backend methods could explicitly trap the error and convert it to the correct target. 

I really like this idea.  How about in addition to the `sqlbk.ErrRetry` we could expose `sqlbk.ErrNotFound` and the like?

Reply:
This is done.